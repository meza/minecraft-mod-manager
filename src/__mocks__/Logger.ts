import { vi } from 'vitest';
import { Logger as Original } from '../lib/Logger.js';

export const Logger = vi.importMock('../lib/Logger.js') as unknown as Original;
