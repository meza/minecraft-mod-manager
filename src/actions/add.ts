import chalk from 'chalk';
import inquirer from 'inquirer';
import path from 'path';
import { CouldNotFindModException } from '../errors/CouldNotFindModException.js';
import { DownloadFailedException } from '../errors/DownloadFailedException.js';
import { NoRemoteFileFound } from '../errors/NoRemoteFileFound.js';
import { UnknownPlatformException } from '../errors/UnknownPlatformException.js';
import { modNotFound } from '../interactions/modNotFound.js';
import { noRemoteFileFound } from '../interactions/noRemoteFileFound.js';
import { ensureConfiguration, getModsFolder, readLockFile, writeConfigFile, writeLockFile } from '../lib/config.js';
import { downloadFile } from '../lib/downloader.js';
import { Logger } from '../lib/Logger.js';
import { Mod, Platform } from '../lib/modlist.types.js';
import { DefaultOptions, telemetry } from '../mmm.js';
import { fetchModDetails } from '../repositories/index.js';

export interface AddOptions extends DefaultOptions {
  allowVersionFallback?: boolean;
  version?: string | undefined;
}

const handleUnknownPlatformException = async (error: UnknownPlatformException, id: string, options: AddOptions, logger: Logger) => {
  const platformUsed = error.platform;
  const platformList = Object.values(Platform);

  if (options.quiet === true) {
    logger.error(`Unknown platform "${chalk.whiteBright(platformUsed)}". Please use one of the following: ${chalk.whiteBright(platformList.join(', '))}`);
  }

  const answers = await inquirer.prompt([
    {
      type: 'list',
      name: 'platform',
      default: false,
      choices: [...platformList, 'cancel'],
      message: chalk.redBright(`The platform you entered (${chalk.whiteBright(platformUsed)}) is not a valid platform.\n`)
        + chalk.whiteBright('Would you like to retry with a valid one?')
    }
  ]);

  if (answers.platform === 'cancel') {
    return;
  }

  // eslint-disable-next-line no-use-before-define
  await add(answers.platform, id, options, logger);

};

export const add = async (platform: Platform, id: string, options: AddOptions, logger: Logger) => {
  performance.mark('add-start');
  const configuration = await ensureConfiguration(options.config, logger, options.quiet);
  const modConfig = configuration.mods.find((mod: Mod) => (mod.id === id && mod.type === platform));

  if (modConfig) {
    logger.debug(`Mod ${id} for ${platform} already exists in the configuration`);
    await telemetry.captureCommand({
      command: 'add',
      success: true,
      arguments: {
        options: options,
        platform: platform,
        id: id
      },
      extra: {
        flag: 'already-exists'
      },
      duration: performance.measure('add-duration', 'add-start', 'add-succeed').duration
    });
    return;
  }

  try {
    const modData = await fetchModDetails(
      platform,
      id,
      configuration.defaultAllowedReleaseTypes,
      configuration.gameVersion,
      configuration.loader,
      !!options.allowVersionFallback,
      options.version
    );

    await downloadFile(modData.downloadUrl, path.resolve(getModsFolder(options.config, configuration), modData.fileName));

    const installations = await readLockFile(options, logger);

    configuration.mods.push({
      type: platform,
      id: id,
      name: modData.name
    });

    installations.push({
      name: modData.name,
      type: platform,
      id: id,
      fileName: modData.fileName,
      releasedOn: modData.releaseDate,
      hash: modData.hash,
      downloadUrl: modData.downloadUrl
    });

    await writeConfigFile(configuration, options, logger);
    await writeLockFile(installations, options, logger);

    logger.log(`${chalk.green('\u2705')} Added ${modData.name} (${id}) for ${platform}`);
    performance.mark('add-succeed');

    await telemetry.captureCommand({
      command: 'add',
      success: true,
      arguments: {
        options: options,
        platform: platform,
        id: id
      },
      duration: performance.measure('add-duration', 'add-start', 'add-succeed').duration
    });

  } catch (error) {
    performance.mark('add-failed');
    await telemetry.captureCommand({
      command: 'add',
      success: false,
      config: configuration,
      arguments: {
        options: options,
        platform: platform,
        id: id
      },
      error: (error as Error).message,
      duration: performance.measure('add-duration', 'add-start', 'add-failed').duration
    });
    if (error instanceof DownloadFailedException) {
      logger.error(error.message, 1);
    }
    if (error instanceof UnknownPlatformException) {
      await handleUnknownPlatformException(error, id, options, logger);
      return;
    }

    if (error instanceof CouldNotFindModException) {
      const { id: newId, platform: newPlatform } = await modNotFound(id, platform, logger, options);
      await add(newPlatform, newId, options, logger);
      return;
    }

    if (error instanceof NoRemoteFileFound) {
      const {
        id: newId,
        platform: newPlatform
      } = await noRemoteFileFound(id, platform, configuration, logger, options);
      await add(newPlatform, newId, options, logger);
      return;
    }

    logger.error((error as Error).message, 2);
  }

};
