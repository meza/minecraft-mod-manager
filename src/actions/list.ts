import chalk from 'chalk';
import { ensureConfiguration, readLockFile } from '../lib/config.js';
import { Logger } from '../lib/Logger.js';
import { Mod } from '../lib/modlist.types.js';
import { DefaultOptions, telemetry } from '../mmm.js';

export type ListOptions = DefaultOptions

export const list = async (options: ListOptions, logger: Logger) => {
  performance.mark('list-start');
  const config = await ensureConfiguration(options.config, logger);
  const installed = await readLockFile(options, logger);

  logger.log((chalk.green('Configured mods')), true);

  const sortByName = (a: Mod, b: Mod) => {
    return a.name.localeCompare(b.name);
  };

  config.mods.sort(sortByName).forEach((mod) => {
    if (installed.find((i) => i.id === mod.id && i.type === mod.type)) {
      logger.log(`${chalk.green('\u2705')} ${mod.name?.trim()} ${chalk.gray('(')}${chalk.gray(mod.id)}${chalk.gray(')')} is installed`, true);
    } else {
      logger.log(`${chalk.red('\u274c')} ${mod.name?.trim()} ${chalk.gray('(')}${chalk.gray(mod.id)}${chalk.gray(')')} is not installed`, true);
    }
  });

  performance.mark('list-succeed');

  await telemetry.captureCommand({
    command: 'list',
    success: true,
    arguments: options,
    extra: {
      numberOfMods: config.mods.length,
      perf: performance.getEntries()
    },
    duration: performance.measure('list-duration', 'list-start', 'list-succeed').duration
  });
};
