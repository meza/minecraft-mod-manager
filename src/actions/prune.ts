import chalk from 'chalk';
import { DefaultOptions, telemetry } from '../mmm.js';
import { Logger } from '../lib/Logger.js';
import { ensureConfiguration, getModsFolder, readLockFile } from '../lib/config.js';
import path from 'node:path';
import fs from 'fs/promises';
import { fileIsManaged } from '../lib/configurationHelper.js';
import { shouldPruneFiles } from '../interactions/shouldPruneFiles.js';
import { getModFiles } from '../lib/fileHelper.js';

export interface PruneOptions extends DefaultOptions {
  force: boolean;
}

export const prune = async (options: PruneOptions, logger: Logger) => {
  performance.mark('prune-start');
  const configuration = await ensureConfiguration(options.config, logger);
  const installations = await readLockFile(options, logger);
  const modsFolder = getModsFolder(options.config, configuration);

  const files = await getModFiles(options.config, configuration);

  if (files.length === 0) {
    logger.log('You have no files in your mods folder.');
    return;
  }

  const unmanaged = files.filter((file) => {
    return !fileIsManaged(file, installations);
  });

  if (unmanaged.length === 0) {
    logger.log('You have no unmanaged mods in your mods folder.');
    return;
  }

  logger.log('The following files are unmanaged:');
  for (const file of unmanaged) {
    logger.log(`${chalk.red('\u274c')} ${file}`);
  }

  if (!await shouldPruneFiles(options, logger)) {
    performance.mark('prune-cancelled');
    await telemetry.captureCommand({
      command: 'prune',
      success: true,
      arguments: {
        options: options
      },
      duration: performance.measure('prune-duration', 'prune-start', 'prune-cancelled').duration
    });
    return;
  }

  for (const file of unmanaged) {
    const filePath = path.resolve(modsFolder, file);
    await fs.rm(filePath, { force: true });
    logger.log(`Deleted: ${filePath}`);
  }

  performance.mark('prune-succeed');

  await telemetry.captureCommand({
    command: 'prune',
    success: true,
    arguments: {
      options: options
    },
    duration: performance.measure('prune-duration', 'prune-start', 'prune-succeed').duration
  });
};
