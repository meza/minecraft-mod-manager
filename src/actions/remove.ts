import { DefaultOptions, telemetry } from '../mmm.js';
import { Logger } from '../lib/Logger.js';
import { ensureConfiguration, getModsFolder, readLockFile, writeConfigFile, writeLockFile } from '../lib/config.js';
import { findLocalMods, getInstallation, hasInstallation } from '../lib/configurationHelper.js';
import fs from 'fs/promises';
import path from 'path';
import chalk from 'chalk';

export interface RemoveOptions extends DefaultOptions {
  dryRun: boolean;
}

export const removeAction = async (mods: string[], options: RemoveOptions, logger: Logger) => {
  performance.mark('remove-start');
  const configuration = await ensureConfiguration(options.config, logger);
  const installations = await readLockFile(options, logger);
  const matches = findLocalMods(mods, configuration);
  const modsDir = getModsFolder(options.config, configuration);

  if (options.dryRun) {
    logger.log(chalk.yellow('Running in dry-run mode. Nothing will actually be removed.'));
  }

  for (const modToBeDeleted of matches) {
    const name = structuredClone(modToBeDeleted.name);
    performance.mark('remove-mod-start', {
      detail: name
    });
    if (options.dryRun) {
      logger.log(`Would have removed ${name}`);
      continue;
    }

    if (hasInstallation(modToBeDeleted, installations)) {
      const installationIndex = getInstallation(modToBeDeleted, installations);
      const filename = installations[installationIndex].fileName;

      if (!options.dryRun) {
        await fs.rm(path.resolve(modsDir, filename), { force: true });
        installations.splice(installationIndex, 1);
        /**
         * We're using structuredClone here to avoid weird reference changes under the hood.
         * This way we write an accurate snapshot of what should be written.
         * The tests have caught a weird race condition that isn't present this way.
         */
        await writeLockFile(structuredClone(installations), options, logger);
      }
    }

    const modIndex = configuration.mods.findIndex((mod) => {
      return (mod.id === modToBeDeleted.id) && (mod.type === modToBeDeleted.type);
    });

    configuration.mods.splice(modIndex, 1);
    await writeConfigFile(configuration, options, logger);

    logger.log(`Removed ${name}`);
    performance.mark('remove-mod-end', {
      detail: name
    });
  }

  performance.mark('remove-succeed');
  await telemetry.captureCommand({
    command: 'remove',
    success: true,
    arguments: {
      options: options,
      mods: mods
    },
    duration: performance.measure('remove-duration', 'remove-start', 'remove-succeed').duration
  });
};
