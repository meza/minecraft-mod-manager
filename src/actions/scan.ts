import { DefaultOptions, telemetry } from '../mmm.js';
import { Logger } from '../lib/Logger.js';
import { ensureConfiguration, getModsFolder, readLockFile, writeConfigFile, writeLockFile } from '../lib/config.js';
import { PlatformLookupResult } from '../repositories/index.js';
import { Mod, ModInstall, ModsJson, Platform, RemoteModDetails } from '../lib/modlist.types.js';
import { scan as scanLib } from '../lib/scan.js';
import chalk from 'chalk';
import { shouldAddScanResults } from '../interactions/shouldAddScanResults.js';

import { getInstallation } from '../lib/configurationHelper.js';
import { getModFiles } from '../lib/fileHelper.js';
import path from 'path';

export interface ScanOptions extends DefaultOptions {
  prefer: Platform;
  add: boolean;
}

export interface ScanResults {
  preferredDetails: RemoteModDetails;
  allRemoteDetails: RemoteModDetails[];
  localDetails: PlatformLookupResult[];
}

export interface FoundEntries {
  mod: Mod;
  install: ModInstall;
}

export interface UnsureEntries {
  configuredMod: number;
  configuredInstallation: number;
  newMod: Mod;
  installation: ModInstall;
}

const findInConfiguration = (platform: Platform, modId: string, configuration: ModsJson) => {
  return configuration.mods.findIndex((configuredMod) => {
    return (configuredMod.type === platform) && (configuredMod.id === modId);
  });
};

const processForeignFiles = async (
  options: ScanOptions,
  configuration: ModsJson,
  dealtWith: string[],
  hasResults: boolean,
  logger: Logger
) => {
  const modsFolder = getModsFolder(options.config, configuration);
  const allFiles = await getModFiles(options.config, configuration);

  const nonMatchedFiles = allFiles.filter((filePath) => {
    const foundIndex = dealtWith.findIndex((dealtWithFile) => {
      return path.resolve(modsFolder, dealtWithFile) === filePath;
    });
    return foundIndex < 0;
  });

  const hasForeignFiles = (nonMatchedFiles.length > 0);

  if ((!hasResults) && (!hasForeignFiles)) {
    logger.log(`${chalk.green('\u2705')} All of your mods are managed by mmm.`);
    return;
  }

  if (!hasResults) {
    logger.log(`${chalk.green('\u2705')} Every mod that can be matched are managed by mmm.`);
  }

  if (hasForeignFiles) {
    logger.log('\nThe following files cannot be matched to any mod on any of the platforms:\n');
    nonMatchedFiles.forEach((file) => {
      logger.log(`  ${chalk.red('\u274c')} ${file}`);
    });
  }
};

export const processScanResults = (scanResults: ScanResults[], configuration: ModsJson, installations: ModInstall[], logger: Logger) => {
  const unmanaged: FoundEntries[] = [];
  const unsure: UnsureEntries[] = [];

  scanResults.forEach((hit) => {

    // ================================================================================================================
    // We know that `hit` has no direct matches in the lockfile, so we need to look for a configuration match

    const halfMatching = hit.localDetails.map((local, index) => {
      const modIndex = findInConfiguration(local.platform, local.modId, configuration);
      if (modIndex < 0) {
        return {
          found: false,
          configured: null,
          remote: hit.allRemoteDetails[index],
          local: local
        };
      }

      return {
        found: true,
        configured: configuration.mods[modIndex],
        remote: hit.allRemoteDetails[index],
        local: local
      };
    }).filter((result) => {
      return result.found === true;
    });

    if (halfMatching.length > 0 && halfMatching[0].configured) {
      const installationIndex = getInstallation(halfMatching[0].configured, installations);

      if (installationIndex > -1) {
        const installation = installations[installationIndex];
        if (halfMatching[0].local.mod.hash !== installation.hash) {
          logger.log(`${chalk.red('\u274c')} ${halfMatching[0].remote.name} has a different version locally than what is in the lockfile`);
        }
      } else {
        logger.log(`${chalk.red('\u274c')} ${halfMatching[0].remote.name} has a local file that isn't in the lockfile.`);
      }

      unsure.push({
        configuredMod: findInConfiguration(halfMatching[0].configured?.type, halfMatching[0].configured?.id, configuration),
        configuredInstallation: installationIndex,
        newMod: {
          type: halfMatching[0].local.platform,
          id: halfMatching[0].local.modId,
          name: halfMatching[0].remote.name
        } as Mod,
        installation: {
          hash: halfMatching[0].local.mod.hash,
          fileName: halfMatching[0].local.mod.fileName,
          name: halfMatching[0].remote.name,
          type: halfMatching[0].local.platform,
          id: halfMatching[0].local.modId,
          releasedOn: halfMatching[0].local.mod.releaseDate,
          downloadUrl: halfMatching[0].local.mod.downloadUrl
        } as ModInstall
      });
    }

    // ================================================================================================================

    if (halfMatching.length === 0) {
      const message = chalk.green('\u2705') + `Found unmanaged mod: ${chalk.bold(chalk.whiteBright(hit.preferredDetails.name))}`;
      logger.log(message, true);

      unmanaged.push(
        {
          mod: {
            type: hit.localDetails[0].platform,
            id: hit.localDetails[0].modId,
            name: hit.preferredDetails.name
          },
          install: {
            name: hit.preferredDetails.name,
            type: hit.localDetails[0].platform,
            id: hit.localDetails[0].modId,
            fileName: hit.localDetails[0].mod.fileName,
            hash: hit.localDetails[0].mod.hash,
            downloadUrl: hit.localDetails[0].mod.downloadUrl,
            releasedOn: hit.localDetails[0].mod.releaseDate
          }
        });
    }
  });
  return {
    unmanaged: unmanaged,
    unsure: unsure
  };
};

export const scan = async (options: ScanOptions, logger: Logger) => {
  performance.mark('scan-start');
  const configuration = await ensureConfiguration(options.config, logger);
  const installations = await readLockFile(options, logger);
  let scanResults: ScanResults[] = [];

  try {
    scanResults = await scanLib(options.config, options.prefer, configuration, installations);
  } catch (error) {
    logger.error((error as Error).message, 2);
  }

  const { unmanaged, unsure } = processScanResults(scanResults, configuration, installations, logger);

  const hasResults = (scanResults.length > 0);
  if (hasResults && await shouldAddScanResults(options, logger)) {

    unmanaged.forEach(({ mod, install }) => {
      configuration.mods.push(mod);
      installations.push(install);
    });

    unsure.forEach((result) => {
      configuration.mods[result.configuredMod] = result.newMod;

      if (result.configuredInstallation > -1) {
        installations[result.configuredInstallation] = result.installation;
      } else {
        installations.push(result.installation);
      }

      logger.log(`${chalk.green('\u2705')} Updated ${configuration.mods[result.configuredMod].name} to match the installed file`);

    });

    await writeConfigFile(configuration, options, logger);
    await writeLockFile(installations, options, logger);
  }

  const dealtWith: string[] = [];

  unsure.forEach((managed) => {
    dealtWith.push(managed.installation.fileName);
  });

  unmanaged.forEach((managed) => {
    dealtWith.push(managed.install.fileName);
  });

  await processForeignFiles(options, configuration, dealtWith, hasResults, logger);

  performance.mark('scan-succeed');

  await telemetry.captureCommand({
    command: 'scan',
    success: true,
    arguments: {
      options: options
    },
    duration: performance.measure('scan-duration', 'scan-start', 'scan-succeed').duration
  });

};
