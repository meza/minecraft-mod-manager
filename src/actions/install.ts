import chalk from 'chalk';
import {
  ensureConfiguration,
  fileExists,
  getModsFolder,
  readLockFile,
  writeConfigFile,
  writeLockFile
} from '../lib/config.js';
import path from 'path';
import { fetchModDetails } from '../repositories/index.js';
import { downloadFile } from '../lib/downloader.js';
import { Mod, ModInstall, ModsJson, Platform, RemoteModDetails } from '../lib/modlist.types.js';
import { getHash } from '../lib/hash.js';
import { DefaultOptions, telemetry } from '../mmm.js';
import { updateMod } from '../lib/updater.js';
import { Logger } from '../lib/Logger.js';
import { fileIsManaged, getInstallation, hasInstallation } from '../lib/configurationHelper.js';
import { handleFetchErrors } from '../errors/handleFetchErrors.js';
import { getModFiles } from '../lib/fileHelper.js';
import { scanFiles } from '../lib/scan.js';
import { processScanResults } from './scan.js';

const getMod = async (moddata: RemoteModDetails, modsFolder: string) => {
  await downloadFile(moddata.downloadUrl, path.resolve(modsFolder, moddata.fileName));
  return {
    fileName: moddata.fileName,
    releasedOn: moddata.releaseDate,
    hash: moddata.hash,
    downloadUrl: moddata.downloadUrl
  };
};

const handleUnknownFiles = async (options: DefaultOptions, configuration: ModsJson, installations: ModInstall[], logger: Logger) => {
  //const modsFolder = getModsDir(options.config, configuration.modsFolder);
  const allFiles = await getModFiles(options.config, configuration);
  const nonManagedFiles = allFiles.filter((filePath) => {
    return !fileIsManaged(filePath, installations);
  });

  if (nonManagedFiles.length === 0) {
    return;
  }

  const scanResults = await scanFiles(nonManagedFiles, installations, Platform.MODRINTH, configuration);
  const { unsure } = processScanResults(scanResults, configuration, installations, logger);

  if (unsure.length > 0) {
    logger.error('\nPlease fix the unresolved issues above manually or by running mmm scan, then try again.', 1);
  }
};

export const install = async (options: DefaultOptions, logger: Logger) => {
  performance.mark('install-start');
  const configuration = await ensureConfiguration(options.config, logger);
  const installations = await readLockFile(options, logger);
  await handleUnknownFiles(options, configuration, installations, logger);
  const installedMods = installations;
  const mods = configuration.mods;
  const modsFolder = getModsFolder(options.config, configuration);

  const processMod = async (mod: Mod, index: number) => {
    const canonVersion = mod.version || 'latest';
    try {
      logger.debug(`Checking ${mod.name}@${canonVersion} for ${mod.type}`);

      if (hasInstallation(mod, installations)) {
        const installedModIndex = getInstallation(mod, installedMods);

        const modPath = path.resolve(getModsFolder(options.config, configuration), installedMods[installedModIndex].fileName);

        if (!await fileExists(modPath)) {
          logger.log(`${mod.name} doesn't exist, downloading from ${installedMods[installedModIndex].type}`);
          await downloadFile(installedMods[installedModIndex].downloadUrl, modPath);
          return;
        }

        const installedHash = await getHash(modPath);
        if (installedMods[installedModIndex].hash !== installedHash) {
          logger.log(`${mod.name} has hash mismatch, downloading from source`);
          await updateMod(installedMods[installedModIndex], modPath, modsFolder);
          return;
        }
        return;
      }

      const modData = await fetchModDetails(
        mod.type,
        mod.id,
        mod.allowedReleaseTypes || configuration.defaultAllowedReleaseTypes,
        configuration.gameVersion,
        configuration.loader,
        !!mod.allowVersionFallback,
        mod.version
      );

      mods[index].name = modData.name;

      // no installation exists
      logger.log(`${mod.name} doesn't exist, downloading from ${mod.type}`);
      const dlData = await getMod(modData, modsFolder);

      installedMods.push({
        name: modData.name,
        type: mod.type,
        id: mod.id,
        fileName: dlData.fileName,
        releasedOn: dlData.releasedOn,
        hash: dlData.hash,
        downloadUrl: dlData.downloadUrl
      });
      return;
    } catch (error) {
      handleFetchErrors(error as Error, mod, logger);
    }

  };

  const promises = mods.map(processMod);

  await Promise.all(promises);

  await writeLockFile(installedMods, options, logger);
  await writeConfigFile(configuration, options, logger);
  logger.log(`${chalk.green('\u2705')} all mods are installed!`);
  performance.mark('install-succeed');

  await telemetry.captureCommand({
    command: 'install',
    success: true,
    arguments: {
      options: options
    },
    duration: performance.measure('install-duration', 'install-start', 'install-succeed').duration
  });
};
