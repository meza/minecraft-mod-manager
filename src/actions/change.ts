import { VerifyUpgradeOptions } from '../lib/verifyUpgrade.js';
import { Logger } from '../lib/Logger.js';
import { telemetry } from '../mmm.js';
import { testGameVersion } from './testGameVersion.js';
import { Mod } from '../lib/modlist.types.js';
import {
  ensureConfiguration,
  fileExists, getModsFolder, readLockFile,
  writeConfigFile,
  writeLockFile
} from '../lib/config.js';
import path from 'path';
import fs from 'node:fs/promises';
import { install } from './install.js';
import { getInstallation, hasInstallation } from '../lib/configurationHelper.js';

export const changeGameVersion = async (gameVersion: string, options: VerifyUpgradeOptions, logger: Logger) => {
  performance.mark('change-start');
  const { version } = await testGameVersion(gameVersion, options, logger);
  performance.mark('change-version-succeed');

  const configuration = await ensureConfiguration(options.config, logger, options.quiet);
  const installations = await readLockFile(options, logger);

  const installedMods = installations;
  const mods = configuration.mods;
  const removeLocalFile = async (mod: Mod) => {

    if (hasInstallation(mod, installations)) {
      const installedModIndex = getInstallation(mod, installedMods);
      const oldModPath = path.resolve(getModsFolder(options.config, configuration), installedMods[installedModIndex].fileName);
      if (await fileExists(oldModPath)) {
        await fs.rm(oldModPath);
      }
    }
  };

  const promises = mods.map(removeLocalFile);

  await Promise.allSettled(promises);

  configuration.gameVersion = version;

  await writeLockFile([], options, logger);
  await writeConfigFile(configuration, options, logger);

  await install(options, logger);

  performance.mark('change-succeed');

  await telemetry.captureCommand({
    command: 'change',
    success: true,
    arguments: {
      options: options,
      gameVersion: gameVersion
    },
    extra: {
      versionTestDuration: performance.measure('change-version-duration', 'change-start', 'change-version-succeed').duration
    },
    duration: performance.measure('change-duration', 'change-start', 'change-succeed').duration
  });

};
