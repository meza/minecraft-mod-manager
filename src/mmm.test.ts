import { chance } from 'jest-chance';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { add } from './actions/add.js';
import { changeGameVersion } from './actions/change.js';
import { install } from './actions/install.js';
import { list } from './actions/list.js';
import { prune } from './actions/prune.js';
import { removeAction } from './actions/remove.js';
import { scan } from './actions/scan.js';
import { testGameVersion } from './actions/testGameVersion.js';
import { update } from './actions/update.js';
import { initializeConfig } from './interactions/initializeConfig.js';
import { Logger } from './lib/Logger.js';
import { Platform } from './lib/modlist.types.js';
import { Telemetry } from './telemetry/telemetry.js';

vi.mock('./telemetry/telemetry.js', () => {
  const telemetryInstance = {
    flush: vi.fn()
  };

  return {
    Telemetry: vi.fn(() => telemetryInstance)
  };
});
vi.mock('./lib/Logger.js');
vi.mock('./actions/add.js');
vi.mock('./actions/list.js');
vi.mock('./actions/scan.js');
vi.mock('./actions/prune.js');
vi.mock('./actions/install.js');
vi.mock('./actions/update.js');
vi.mock('./interactions/initializeConfig.js');
vi.mock('./actions/testGameVersion.js');
vi.mock('./actions/change.js');
vi.mock('./actions/remove.js');

describe('The main CLI configuration', () => {
  let logger: Logger;
  let telemetryInstance: Telemetry;
  beforeEach(() => {
    vi.resetAllMocks();
    vi.resetModules();
    telemetryInstance = {
      flush: vi.fn()
    } as unknown as Telemetry;
    vi.mocked(Telemetry).mockImplementation(() => telemetryInstance);
    vi.spyOn(process, 'cwd').mockReturnValue('/path/to/minecraft/installation');
    logger = new Logger({} as never);
  });

  it('is set up correctly', async () => {
    const { program } = await import('./mmm.js');
    expect(program.helpInformation()).toMatchSnapshot();
  });

  it('has add hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(add).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['add', 'a']),
      chance.pickone(Object.values(Platform)),
      chance.word()]);
    expect(vi.mocked(add)).toHaveBeenCalledOnce();
  });

  it('has list hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(list).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['list', 'ls'])
    ]);
    expect(vi.mocked(list)).toHaveBeenCalledOnce();
  });

  it('has install hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(install).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['install', 'i'])
    ]);
    expect(vi.mocked(install)).toHaveBeenCalledOnce();
  });

  it('has update hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(update).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['update', 'u'])
    ]);
    expect(vi.mocked(update)).toHaveBeenCalledOnce();
  });

  it('has initialize hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(initializeConfig).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['init'])
    ]);
    expect(vi.mocked(initializeConfig)).toHaveBeenCalledOnce();
  });

  it('has the test hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(testGameVersion).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['test', 't'])
    ]);
    expect(testGameVersion).toHaveBeenCalledOnce();
  });

  it('has the change hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(changeGameVersion).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['change'])
    ]);
    expect(changeGameVersion).toHaveBeenCalledOnce();
  });

  it('has the scan hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(scan).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['scan'])
    ]);
    expect(scan).toHaveBeenCalledOnce();
  });

  it('has the prune hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(prune).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['prune'])
    ]);
    expect(prune).toHaveBeenCalledOnce();
  });

  it('has the remove action hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(removeAction).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['remove']),
      []
    ]);
    expect(removeAction).toHaveBeenCalledOnce();
  });

  it('sets the logger to quiet when the quiet option is supplied', async () => {
    const { program } = await import('./mmm.js');
    await program.parse([
      '',
      '',
      chance.pickone(['-q', '--quiet']),
      chance.pickone(['init'])
    ]);
    expect(logger.flagQuiet).toHaveBeenCalledOnce();
  });

  it('sets the logger to debug when the debug option is supplied', async () => {
    const { program } = await import('./mmm.js');
    await program.parse([
      '',
      '',
      chance.pickone(['-d', '--debug']),
      chance.pickone(['init'])
    ]);
    expect(logger.flagDebug).toHaveBeenCalledOnce();
  });

  it('can stop the execution', async () => {
    vi.spyOn(process, 'exit').mockImplementation(() => {
      throw new Error('process.exit');
    });
    const { stop, telemetry } = await import('./mmm.js');

    await expect(stop).rejects.toThrow('process.exit');
    expect(telemetry.flush).toHaveBeenCalledOnce();
  });
});
