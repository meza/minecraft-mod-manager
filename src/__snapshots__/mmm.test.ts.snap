// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The main CLI configuration > is set up correctly 1`] = `
"Usage: Minecraft Mod Manager [options] [command]

Manages mods from Modrinth and Curseforge

Options:
  -V, --version                    output the version number
  -c, --config <MODLIST_JSON>      An alternative JSON file containing the configuration (default: "./modlist.json")
  -q, --quiet                      Suppress all output (default: false)
  -d, --debug                      Enable debug messages (default: false)
  -h, --help                       display help for command

Commands:
  list|l
  install|i
  update|u
  add|a [options] <type> <id>
  init [options]
  test|t [game_version]
  change [options] [game_version]
  scan [options]                   Scans the mod directory and attempts to find the mods on the supported mod platforms.
  prune [options]                  Prunes the mod directory from all the unmanaged files.
  remove [options] <mods...>       Removes one or more mods from both the config and the filesystem.
  help [command]                   display help for command
"
`;
